Makefile update

marc1uk opened the pull request:
Update Makefile to enable building only tools associated to a specific toolchain. Use `make <ToolChainName>` to build only tools associated to that toolchain, potentially speeding up compilation and link time. Using just `make` will preserve old behaviour of building everything.
Some files shared across multiple Tools but not in the DataModel needed to be moved. Some Tool files were changed as fixes to the makefile required them (clash of global constants)

## Checklist before submitting your PR
- [✓] This PR implements a single change (one new/modified Tool, or a set of changes to implement one new/modified feature)
- [✓] This PR alters the minimum number of files to affect this change
- [N/A] If this PR includes a new Tool, a README and minimal demonstration ToolChain is provided
- [N/A] If a new Tool/ToolChain requires model or configuration files, their paths are not hard-coded, and means of generating those files is described in the readme, with examples provided on /pnfs/annie/persistent
- [N/A] For every `new` usage, there is a reason the data must be on the heap
- [N/A] For every `new` there is a `delete`, unless I explicitly know why (e.g. ROOT or a BoostStore takes ownership)

Commit:
update makefile to allow building a specific toolchain. Associated changes to support this; move classes shared by multiple tools to DataModel, update PlotWaveforms library Makefile for new locations, change pre-processor defines duplicated in multiple headers to consts in implemenetation file.

Commit:
make MyFactory if it doesn't exist

jminock commented:
Hi Marcus, immediate things that stick out is the build fails with the following error:
`In file included from UserTools/MyFactory/Unity.h:154,
                 from UserTools/MyFactory/Factory.h:6,
                 from UserTools/MyFactory/MyFactory.cpp:1:
include/parseLAPPDData.h:15:25: error: expected unqualified-id before numeric constant
 #define NUM_VECTOR_DATA 7795
                         ^~~~
include/parseLAPPDData.h:15:25: note: in definition of macro 'NUM_VECTOR_DATA'
 #define NUM_VECTOR_DATA 7795
                         ^~~~
compilation terminated due to -fmax-errors=1.
make: [Makefile:150: UserTools/MyFactory/MyFactory.o] Error 1 (ignored)`

Also, could you please create documentation for how to compile specific toolchains in some form of README in ToolAnalysis? Given this adds functionality to ToolAnalysis, it would be great for there to be a reference within ToolAnalysis or with other ToolAnalysis usage documentation. Thanks!

superkanalysis committed:
missed Tool define collisions. Note these same defines are made in other Tools so should be moved to DataModel/Constants.h

superkanalysis committed:
add notes on building to README

marc1uk commented:
OK, not sure why it built ok on the warwick test stand but not the CI - i guess that's why we have CI runners! Fixed the issue and added some comments to the ToolAnalysis README.

jminock commented on Makefile:
```diff
@@ -58,10 +70,10 @@ MyToolsInclude += `python3-config --cflags` -Wno-sign-compare
-all: lib/libStore.so lib/libLogging.so lib/libDataModel.so include/Tool.h lib/libMyTools.so lib/libServiceDiscovery.so lib/libToolChain.so Analyse
-
-Analyse: src/main.cpp | lib/libMyTools.so lib/libStore.so lib/libLogging.so lib/libToolChain.so lib/libDataModel.so lib/libServiceDiscovery.so
+Analyse: src/main.cpp | lib/libStore.so lib/libLogging.so lib/libToolChain.so lib/libServiceDiscovery.so lib/libMyTools.so lib/libToolChain.so lib/libDataModel.so lib/libServiceDiscovery.so
```
Any reason why `lib/libToolChain.so`  and `lib/libServiceDiscovery.so` are here twice?

marc1uk replied:
nope, removed duplicates :+1:

jminock replied:
Thank you! I'll wait on the build to finish before I review again

jminock commented on Makefile:
```diff
@@ -70,26 +82,23 @@ lib/libStore.so: $(ToolDAQPath)/ToolDAQFramework/src/Store/*
 
-
-lib/libToolChain.so: $(ToolDAQPath)/ToolDAQFramework/src/ToolChain/* | lib/libLogging.so lib/libStore.so lib/libServiceDiscovery.so lib/libLogging.so lib/libDataModel.so
+lib/libToolChain.so: $(ToolDAQPath)/ToolDAQFramework/src/ToolChain/* UserTools/MyFactory/Unity.h | lib/libLogging.so lib/libStore.so lib/libServiceDiscovery.so lib/libLogging.so lib/libDataModel.so
```
Also a repeat here for `lib/libLogging.so`

jminock reviewed (commented):
Overall looks good. My only comment is about the repeats in the Makefile. If it builds fine, then I don't think there's a problem. Would these pose a problem? If not, I'll update this PR with my approval.

superkanalysis committed:
remove duplicate libraries in target dependencies

jminock reviewed (approved):
Looks good! Thank you!!
